components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1305 @@
+        RegionHeartbeatResponse, ReportBucketsRequest, TsoRequest, TsoResponse,
+    },
+    replication_modepb::{ReplicationStatus, StoreDrAutoSyncStatus},
+};
+use security::SecurityManager;
+use tikv_util::{
+    box_err, error, info,
+    mpsc::future as mpsc,
+    slow_log, thd_name,
+    time::{duration_to_sec, Instant},
+    timer::GLOBAL_TIMER_HANDLE,
+    warn,
+};
+use tokio::sync::{Mutex, Notify};
+use txn_types::TimeStamp;
+
+use super::{
+    client::{CLIENT_PREFIX, CQ_COUNT},
+    metrics::*,
+    util::{check_resp_header, PdConnector, TargetInfo},
+    Config, Error, FeatureGate, RegionInfo, Result, UnixSecs,
+    REQUEST_TIMEOUT as REQUEST_TIMEOUT_SEC,
+};
+use crate::PdFuture;
+
+const REQUEST_TIMEOUT: Duration = Duration::from_secs(REQUEST_TIMEOUT_SEC);
+
+/// Immutable context for making new connections.
+struct ConnectContext {
+    cfg: Config,
+    connector: PdConnector,
+}
+
+#[derive(Clone)]
+struct RawClient {
+    stub: PdClientStub,
+    target_info: TargetInfo,
+    members: GetMembersResponse,
+}
+
+impl RawClient {
+    /// Returns Ok(true) when a new connection is established.
+    async fn maybe_reconnect(&mut self, ctx: &ConnectContext, force: bool) -> Result<bool> {
+        PD_RECONNECT_COUNTER_VEC.with_label_values(&["try"]).inc();
+        let start = Instant::now();
+
+        let members = self.members.clone();
+        let direct_connected = self.target_info.direct_connected();
+        slow_log!(start.saturating_elapsed(), "try reconnect pd");
+        let (stub, target_info, members, _) = match ctx
+            .connector
+            .reconnect_pd(
+                members,
+                direct_connected,
+                force,
+                ctx.cfg.enable_forwarding,
+                false,
+            )
+            .await
+        {
+            Err(e) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["failure"])
+                    .inc();
+                return Err(e);
+            }
+            Ok(None) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["no-need"])
+                    .inc();
+                return Ok(false);
+            }
+            Ok(Some(tuple)) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["success"])
+                    .inc();
+                tuple
+            }
+        };
+
+        fail_point!("pd_client_v2_reconnect", |_| Ok(true));
+
+        self.stub = stub;
+        self.target_info = target_info;
+        self.members = members;
+
+        info!("trying to update PD client done"; "spend" => ?start.saturating_elapsed());
+        Ok(true)
+    }
+}
+
+/// A shared [`RawClient`] with a local copy of cache.
+#[derive(Clone)]
+pub struct CachedRawClient {
+    context: Arc<ConnectContext>,
+
+    latest: Arc<Mutex<Option<RawClient>>>,
+    version: Arc<AtomicU64>,
+    on_reconnect_notify: Arc<Notify>,
```

Comment:
Why use `Notify`?

```
    println!("sending notification");
    notify.notify_waiters();

    notify.notified().await;
    println!("received notification");
```

Will only print one line.

Reply:
You are right. Maybe a full-fledged condvar is needed after all. The broadcast channel you proposed earlier isn't very handy either, because its sender might be blocked by an idle receiver that doesn't attempt to consume the messages. A combination of tokio::watch and tokio::broadcast would be perfect here but it doesn't exist...

Reply:
>  its sender might be blocked by an idle receiver 

Why? Isn't it just overwrite the oldest messages?

Reply:
I see.